packages/web/lib/page.js

```diff
@@ -40,15 +99,65 @@ const getPageScope = (parent) => ({
       });
       return null;
     },
+    "@string": dataStorageFn(getString, "string"),
+    "@string?": dataStorageFn(getString, "string", true),
+    "@number": dataStorageFn(getNumber, "number"),
+    "@number?": dataStorageFn(getNumber, "number", true),
+    "@url": dataStorageFn(getUrl, "url"),
+    "@url?": dataStorageFn(getUrl, "url", true),
     click: (scope) => async (selector) => {
+      strictValidateScope(scope, `Cannot call click outside of page context`);
       const page = resolveInternal("page", scope);
       if (!page) {
         return false;
       }
       await page.click(selector);
       return true;
     },
+    config: (scope) => async (ruleSet) => {
+      strictValidateScope(scope, `Cannot call config outside of page context`);
+      //Since config applies to pages, this should set the config for the nearest page
+      const nearestPageScope = resolveInternalScope("page", scope);
+
+      const configScope = getNewScope(nearestPageScope);
+
+      //Override the set behavior
+      configScope.fns.set = (scope) => (name, value) => {
+        //Note: a benefit of validate scope, there is no need to check nearestPageScope, we know we're in a page scope
+        nearestPageScope.internal.config[name] = value;
+        if (name === "output") {
+          //For output files, we create the file and any directories and then open up a write stream
+          fs.ensureFileSync(value);
+          nearestPageScope.internal.config.writeStream = fs.createWriteStream(
+            value
+          );
+        }
+      };
+      //Evaluate the ruleSet
+      await evalRuleSet(ruleSet, configScope);
+
+      return nearestPageScope.internal.config;
+    },
+    crawl: (scope) => async (selector) => {
+      strictValidateScope(scope, `Cannot call crawl outside of page context`);
```

Comment:
`context` -> `scope`?

tbh I've been using `context` and `scope` and `runtime` pretty interchangeably too

We should try and define this terminology (on the wiki) later and go through all our messaging to try and stay consistent

Cleanup work for later